File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.java
class Solution {
    public int minDeletions(String s) {
        int deletions = 0;
        int[] freq = new int[26];
        for(char c: s.toCharArray()){
            freq[c - 'a']++;
        }
        
        Set<Integer> usedFreq = new HashSet<>();
        for(int i: freq){
            if(i != 0){
                boolean added = usedFreq.add(i);
                // If frequency is repeated
                if(!added){
                    // Find previous spot not taken or zero
                    int nextEmpty = i - 1;
                    while(usedFreq.contains(nextEmpty) && nextEmpty != 0){
                        nextEmpty--;
                    }
                    deletions += i - nextEmpty;
                    usedFreq.add(nextEmpty);
                }
            }
        }
        return deletions;
    }
}

File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.js
/**
 * @param {string} s
 * @return {number}
 */
var minDeletions = function(s) {
    var deletions = 0;
    var freq = new Array(26).fill(0);
    const charCodeA = 97;
    for (var i = 0; i < s.length; i++) {
      freq[s.charCodeAt(i) - charCodeA]++;
    }
    
    var used = new Set();
    freq.forEach((i) => {
        if(i !== 0){
            let tempSize = used.size;
            used.add(i);
            let added = tempSize < used.size;
            if(!added){
                let nextEmpty = i - 1;
                while(used.has(nextEmpty) && nextEmpty !== 0){
                    nextEmpty--;
                }
                deletions += i - nextEmpty;
                used.add(nextEmpty);
            }
        }
    });
    return deletions;
};